## Machine Learning - Tasks 2020

### Task 1 - Square Root

Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places. Your code should not depend on any module from the standard library1 or otherwise. You should research the task first and include references and a description of your algorithm.


#### Square Root

In mathematics, a square root of a number x is a number y such that y2 = x; in other words, a number y whose square (the result of multiplying the number by itself, or y ⋅ y) is x.[1.1]

#### Methods of computing Square roots

Methods of computing square roots are numerical analysis algorithms for finding the principal, or non-negative, square root (usually denoted √S, 2√S, or S1/2) of a real number. Arithmetically, it means given S, a procedure for finding a number which when multiplied by itself, yields S; algebraically, it means a procedure for finding the non-negative root of the equation x2 - S = 0; geometrically, it means given the area of a square, a procedure for constructing a side of the square.

Every real number has two square roots.In addition to the principal square root, there is a negative square root equal in magnitude but opposite in sign to the principal square root, except for zero, which has double square roots of zero. The principal square root of most numbers is an irrational number with an infinite decimal expansion. As a result, the decimal expansion of any such square root can only be computed to some finite-precision approximation. However, even if we are taking the square root of a perfect square integer, so that the result does have an exact finite representation, the procedure used to compute it may only return a series of increasingly accurate approximations.

The most common analytical methods are iterative and consist of two steps: finding a suitable starting value, followed by iterative refinement until some termination criteria is met. The starting value can be any number, but fewer iterations will be required the closer it is to the final result. The most familiar such method, most suited for programmatic calculation, is Newton's method, which is based on a property of the derivative in the calculus. 

#### Newton's method 
Newton’s method, also known as Newton-Raphson method is a root-finding algorithm that produces successively better approximations of the roots of a real-valued function. The approximations of the root go as:
x_(n+1) = x_n - f(x_n) / f’(x_n)
x_0 is the rough approximation of the root done at the first and the successive approximations go as x_1, x_2, ….
f(x_n) is the function whose root is to be determined and f’(x_n) is the derivative of the function.[1.3]


In [1]:
# Function to calculate square root using Newton't method
# Sample code taken from https://www.instructables.com/Python-Programming-calculating-Newtons-Method/

# n - used to get input from user
# DP - used to define the number of decimal places
def NewtonSqRoot(n, DP):
    
    #set the loop counter
    i = 1
    x = n
    
    # Set the initial guess for the square root 
    y = (x + 1) / 2
    
    # loop will run until the new guess is less than the previous guess
    while y < x:
        x = y
        y = (x + (n / x)) / 2
        
        print(i, "={:.{}f}".format(x, DP))
        i += 1
    

In [2]:
NewtonSqRoot(2, 100)


1 =1.5000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
2 =1.4166666666666665186369300499791279435157775878906250000000000000000000000000000000000000000000000000
3 =1.4142156862745096645994635764509439468383789062500000000000000000000000000000000000000000000000000000
4 =1.4142135623746898698271934335934929549694061279296875000000000000000000000000000000000000000000000000
5 =1.4142135623730949234300169337075203657150268554687500000000000000000000000000000000000000000000000000


#### Babylonian method
One of the algorithms used for approximating square roof of S is the Babylonian method, despite there being no direct evidence, beyond informed conjecture, that the eponymous Babylonian mathematicians employed exactly this method.[1.1] The basic idea is that if x is an overestimate to the square root of a non-negative real number S then S/x will be an underestimate, or vice versa, and so the average of these two numbers may reasonably be expected to provide a better approximation. This is equivalent to using Newton's method to solve x^2-S=0.

In [3]:
# function to calculate square root using Babylonian method
# Sample code taken from https://www.geeksforgeeks.org/square-root-of-a-perfect-square/

def BabyloniansquareRoot(n, DP): 
    x = n; 
    # sets the initial guess as 1
    y = 1; 
    
    # set Loop counter
    i = 1
    
    # Loop will continue until the new guess is smaller than the previous guess
    while(x > y): 

        x = (x + y) / 2; 
        y = n / x; 
        print(i,"={:.{}f}".format(x, DP));
        i += 1




In [4]:
BabyloniansquareRoot(2,100)

1 =1.5000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
2 =1.4166666666666665186369300499791279435157775878906250000000000000000000000000000000000000000000000000
3 =1.4142156862745096645994635764509439468383789062500000000000000000000000000000000000000000000000000000
4 =1.4142135623746898698271934335934929549694061279296875000000000000000000000000000000000000000000000000
5 =1.4142135623730949234300169337075203657150268554687500000000000000000000000000000000000000000000000000


#### Digit by Digit Method
This is a method to find each digit of the square root in a sequence. It is slower than the Babylonian method, but it has several advantages:

- It can be easier for manual calculations.
- Every digit of the root found is known to be correct, i.e., it does not have to be changed later.
- If the square root has an expansion that terminates, the algorithm terminates after the last digit is found. Thus, it can be used to check whether a given integer is a square number.
- The algorithm works for any base, and naturally, the way it proceeds depends on the base chosen.
Napier's bones include an aid for the execution of this algorithm. The shifting nth root algorithm is a generalization of this method. [1.5]





In [5]:
# function to calculate square root using Digit by Digit method
# from http://stackoverflow.com/questions/15390807/integer-square-root-in-python
def isqrt(n):
    
    x = n
    # Initial guess for the largest integer square root
    # equals slightly more than half of n to make sure the next guess progresses in the right
    # direction when near n is near 1 and avoids halving even numbers
    y = (x + 1) // 2
    # loop until new guess is larger than previous guess 
    while y < x:
        x = y
        #reduce the next guess integer until the next guess is larger than the previous guess
        y = (x + n // x) // 2
    # return the largest guess as result
    return x


def DigitSquareRoot(n,DP):
# n is a variable to store the number we want to get the square root of.

    # Declare the output string
    Strresult = ""
    
    #s = "0"
    # when the given number is a float separate the whole number  and decimal points in to two lists
    # code adapted from https://www.w3schools.com/python/ref_string_split.asp
    if type(n) != int:
        #When the given number is float then we can separate left and right hand array by spliting it with the decimal point
        strSplit = str(n).split(".")
    else:
        # When the given number is not float we can add the decimal point with two zeros to produce array for right hand side
        s = str(n) + ".00"
        strSplit = str(s).split(".")

    #Now arranging the splited numbers into a list    
    LeftArray = [int(x) for x in strSplit[0]]
    RightArray = [int(x) for x in strSplit[1]]

    # Adding zeros to make the left array with minimum two digits        
    if len(LeftArray)%2==1:
        LeftArray.insert(0,0)
        
    # Adding zeros to the Right hand array based on the requested decimal point 
    if len(RightArray)<2*DP:
        for i in range(2*DP-len(RightArray)):
            RightArray.append(0)
            
    if len(RightArray)%2==1:
        RightArray.append(0)
        

    # Below makes the pairs of digits required in the algorithm
    pairs=[[10*LeftArray[2*i]+LeftArray[2*i+1] for i in range(int(len(LeftArray)/2))],[10*RightArray[2*i]+RightArray[2*i+1] for i in range(int(len(RightArray)/2))]]

    # part of the square root of n already found - setting default value
    p=0
    # remainder from previous calculation - setting default value
    r=0
    # setting the default list for the square root list
    square_root=[[],[],"+"]
    for i in range(len(pairs[0])):
        # Extract next paired digit list and add remainder
        # from previous run which is 100 times larger scale
        c = 100 * r + pairs[0][i]
        # Calculate next integer digit of the square root
        x = (-20 * p + isqrt(400 * p ** 2 + 4 * c)) // 2
        # Calculate remainder r 
        r = c - (20 * p * x + x ** 2)
        # Add the next digit to the known part of the square root
        p = 10 * p + x
        # Append x  to the result list
        square_root[0].append(x)

    for i in range(len(pairs[1])):
        # Extract next paired digit list and add remainder
        # from previous run which is 100 times larger scale
        c = 100 * r + pairs[1][i]
        # Calculate next integer digit of the square root
        x = (-20 * p + isqrt(400 * p ** 2 + 4 * c)) // 2
        # Calculate remainder r 
        r = c - (20 * p * x + x ** 2)
        # Add the next digit to the known part of the square root
        p = 10 * p + x
        # Append x  to the result list
        square_root[1].append(x)

    #concatinate the list values to get the final result
    #https://www.techbeamers.com/python-convert-list-string/
    Strresult1 = ''.join( str(x) for x in square_root[0])
    Strresult2 = ''.join( str(x) for x in square_root[1])
    Strresult=Strresult1+'.'+Strresult2
    print(f"The square root of '{n}' using Digit by Digit method with '{DP}' decimal places :")
    print(Strresult)
    

#print(DigitSquareRoot(2,100))

In [6]:
#Compare the results from the above methods
print("Newton Method")
NewtonSqRoot(2, 100)

print("")
print("Babylonian Method")
BabyloniansquareRoot(2,100)

print("")
print("Digit by Digit Method")
DigitSquareRoot(2,100)

Newton Method
1 =1.5000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
2 =1.4166666666666665186369300499791279435157775878906250000000000000000000000000000000000000000000000000
3 =1.4142156862745096645994635764509439468383789062500000000000000000000000000000000000000000000000000000
4 =1.4142135623746898698271934335934929549694061279296875000000000000000000000000000000000000000000000000
5 =1.4142135623730949234300169337075203657150268554687500000000000000000000000000000000000000000000000000

Babylonian Method
1 =1.5000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
2 =1.4166666666666665186369300499791279435157775878906250000000000000000000000000000000000000000000000000
3 =1.4142156862745096645994635764509439468383789062500000000000000000000000000000000000000000000000000000
4 =1.4142135623746898698271934335934929549694061279296875000000000000000000000000000000000000000000000000
5 =1.41421356

#### References:

[1.1] Square Root Definition; https://en.wikipedia.org/wiki/Square_root
[1.2] Methods of calculating Square root; https://en.wikipedia.org/wiki/Methods_of_computing_square_roots
[1.3] Newton't Method; https://surajregmi.medium.com/how-to-calculate-the-square-root-of-a-number-newton-raphson-method-f8007714f64
[1.4] Math functions Basic Square root; https://www.mathsisfun.com/square-root.html
[1.5] Digit by Digit method of Square root calculation ; https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Digit-by-digit_calculation


### Task 2: : Chi-squared Test for Independence

#### Task Description: 
Using scipy.stats verify the Chi-squared test value of ~24.6 for the given table example and calculate the associated independence test p value. Justify the analysis of test.

##### Research
##### Chi-squared test

The chi-squared test procedure can be used to find the significant association between two or more set of data by comparing the actual data to an expected set of data [2.1].Theory of Hypotheses states that it can be of Null or alternative. Null Hypotheses- where the data sets are independent and Alternative Hypotheses- data sets are NOT independent [2.2].


Chi-squared test procedure should satisfy the below mandatory conditions:[2.2]
- The sampling method is simple random sampling
- The variables are categorised
- The sample data set are presented in a table format, minimum expectation to have at least 5 frequency count.


Example of chi-squared test for independence
In the example [4] from the Wikipedia article [2.1] explains the Null Hypotheses- Analyse whether two category variables (Data sets) are independent? Here we have used Variable 1 as residents in a city with four neighbourhoods: A, B, C, D, and Variable 2 as occupation logged based on various collar colours like WHITE, BLUE, NO COLLAR.


![Chi Square Example](images/Chi_Square_example.jpg)

From the above table, the two categorical variables variable 1-Neighborhood with possible values (A,B,C,D) and variable 2 - Occupation with possible values (White Collar, Blue Collar, No Collar). The numbers is each cell contain the observer frequencies, corresponding to the intersection of the two variables. We will use Che-Squared test of independence to check for a statistically significant relationship between the two variables residents and occupation. For the above table there are only random sample of 650 residents of the city has been taken and the total numbe rof residence in the city could be more than a Million. 


The Chi Square statistic is commonly used for testing realtionships between categorical variables. The null hypothesis of the Chi-Square test is that no relationship exists on the categorical variables in the poplulation. Calculating the Chi-Square statistic and comparing it against a critical value from the Chi-Square distribution allows the researcher to assess whether the observed cell counts are significantly different from the expected cell counts.[2.3]


##### Procedure to perform Chi-Squared test
#####  ⦁ Step 1. Calculate the chi square value
Calculate "Expected Value" for each entry: Multiply each row total by each column total and divide by the overall total:


![Step1](images/Chi_square_1.jpg)


which gives us

![Chi_Square2](images/Chi_square_2.jpg)

The formula for the test statistic is: (O-E)^2 / E 
O = Observed (actual) value, E = Expected value
For neighbourhood A, white color the value comes from (90-80.54)2 /80.54 = 1.11
Which gives us:

![Chi_Square2](images/Chi_square_3.jpg)

As per the above table, the test statistic is calculated for each cell and the sum over all cells is the test statistic. Now add up those calculated values to calculate Chi-Square = 24.57

The test statistic has approximately a Chi-Squared distribution whose number of degrees of freedom is calculated below.

##### . Step 2. Calculate the Degree of Freedom

Degree of Freedom = (rows − 1) × (columns − 1)
For our example we have 2 rows and 2 columns: DoF = (3 − 1) × (4 − 1) = 2×3 = 6

##### . Step 3. Calculate the P value.
P value calculated through chi – square test calculator [2.4]
P-value (p) = 0.000404855. 

The measured test statistic is evaluated by comparing it against some critical value in the Che-Squared distribution. The Degree of Freedom and the test statistic are combined to calculate p-value as above and the value is p-0.000404855. The usual test for significance is a p-value <0.5: if p<0.05, the result is significant, we reject the null hypothesis and we cannot assume that the categorical variables are independent. It is possible to chose a lower significance level, meaning that you require more evidence ot reject the null hypothesis.

### Test Chi-Square using SciPy
#### Import required packages

In [7]:
# For statistics
import scipy.stats as ss
# For plotting
import matplotlib.pyplot as plt
plt.rcParams['figure.figsize'] = (10.0, 8.0) # Make nice big figures

In [8]:
# Create the Chi-Square test table from the question
# Code taken from https://machinelearningmastery.com/critical-values-for-statistical-hypothesis-testing/

white = [90, 60, 104, 95] # Array containing Row 1 of the table
blue = [30, 50, 51, 20] # Row 2 of the table
none = [30, 40, 45, 35] # Row 3 of the table
Chitable = [white, blue, none]
print('Chi-Square Sample Table')
print(Chitable)

# Calculate test statistic, p-value, degrees of freedom, and expected values.
stat, p, dof, expected = ss.chi2_contingency(Chitable)
print('\nTest statistic = %.3f' % stat)
print('Degrees of freedom = %d' % dof)
print(f'p-value = {p:.5f}')
print('\nTable of expected frequency values:')
print(expected)

# Chose a significance level: 10%
prob = 0.90

# Percent point function (inverse of cdf — percentiles).
# Given our chosen significance level (10%), calculate the critical value of the test statistic
# using the ppf of a chi2 distribution.
critical = ss.chi2.ppf(prob, dof)
print('\nprobability = %.3f, critical = %.3f, stat = %.3f' % (prob, critical, stat))

Chi-Square Sample Table
[[90, 60, 104, 95], [30, 50, 51, 20], [30, 40, 45, 35]]

Test statistic = 24.571
Degrees of freedom = 6
p-value = 0.00041

Table of expected frequency values:
[[ 80.53846154  80.53846154 107.38461538  80.53846154]
 [ 34.84615385  34.84615385  46.46153846  34.84615385]
 [ 34.61538462  34.61538462  46.15384615  34.61538462]]

probability = 0.900, critical = 10.645, stat = 24.571


from the chosen 10% significance level, p-value is calculated as 0.00041 for our hypothesis test. Since the p-value is <0.05, we must reject the null hypothesis and we cannot say that there is a relationship between Occupation and Neighborhood. 
In terms of crtical values, we have calculated the critical value of 10.645 and the test statistic of 24.571. As out calculated test statistic is > the critical value, we must reject the null hypothesis. The critical value is an alternative method of p-value to test the same. 

So based on the above calculation and analysis, the value of the Chi-Squared test statistic for the provided contingency table is 24.571. This corresponds to a p-value of 0.00041 at a significance level of 10%. So we can reject the null hypothesis that the categorical  variables are not dependent. 


References :
 - [2.1] Chi-square test; Wikipedia; https://en.wikipedia.org/w/index.php?title=Chi-squaredtest&oldid=983024096
 - [2.2] Chi-Square Test for Independence; Stat Trek.com; https://stattrek.com/chi-square-test/independence.aspx
 - [2.3] Using Chi-Square Statistic in Research, https://www.statisticssolutions.com/using-chi-square-statistic-in-research/
 - [2.4] Chi-Square Test, Math is Fun, https://www.mathsisfun.com/data/chi-square-calculator.html
 - [2.5] Scipy References, https://docs.scipy.org/doc/scipy/reference/stats.html

### Task 3: Microsoft Excel methods for Standard Deviation
##### Task Description: 
Research, review and compare two Microsoft Excel functions for Standard Deviation STDEV.P and STDEV.S. Use python numpy simulation to demonstrate that the STDEV.S calculation function is a better estimate for standard deviation of a population when performed on a sample.


##### Research
###### Standard Deviation
The standard deviation measures the distribution of a dataset relative to its mean and is calculated as the square root of the variance. The symbol for standard deviation is σ (sigma) calculated as the square root of averaged square differences from the Mean [3.1]. There are two standard deviations - Population standard deviation (1) and Sample standard deviation (2), which are calculated differently according to the dataset [3.2, 3.3]


In [1]:
!(images/Task3.png)


'images' is not recognized as an internal or external command,
operable program or batch file.


##### When to use the sample or population standard deviation
Population standard deviation (1) are often used when all the values refers to the entire population interested in and sample standard deviation (2) are used when values refers to the sample. Sometimes confusion can often occur to find which standard deviation to use for above cases, but it depends on what outcome is expected at the end result. Refer below example to understand the different standard deviation cases. [3.2]
Example: To measure existing consumer opinion on a product/service, use population standard deviation to provide more quantifiable reliable number. Nevertheless in case of new experiment to attract new consumer, then sample standard deviation is a best choice, because sample result can be categorized based on gender, age and locations.


##### Microsoft Excel STDEV.P function
To calculate the standard deviation for an entire population data, Microsoft excel provides STDEV.P function. The results are accurate because all data are available. Also in some cases of sample data, and only want standard deviation for the sample, without generalising for the entire population, even in those cases STDEV.P function can be used [3.4]

##### Microsoft Excel STDEV.S function
To calculate the standard deviation for a sample data, STDEV.S function can be used. The results are estimates and therefore not as accurate. The STDEV.S function uses Bessel’s corrections to work on sample data to provide better estimates of the standard deviation. Here the Bessel’s corrections appears in the formula as n-1, where n is the count [3.4].

###### Constraints:
 - Data is completely numeric.
 - Empty cells, logical values, text, or error values in the array or reference are ignored [3.5].
 - The deviation for a single value is zero[3.6]
 
###### Procedure to perform standard deviation methods
 - Analyse of the data to be estimated using available standard deviation methods.
 - Analyse and compare the STDEV.P method and STDEV.S method
 - Evaluate Standard Deviation of a population while performed on a sample



References:

[3.1] Standard Deviation and Variance; Mathsisfun; https://www.mathsisfun.Com/data/ standard-deviation.html
[3.2] Standard Deviation; Laerd statistics; https://statistics.laerd.com/statistical-guides/measu res-of-spread-standard-deviation.php
[3.3]Population and sample standard deviation review; khanacademy; https://www .khanacademy.org/math/statistics-probability/summarizing-quantitative-data/variance-standard-deviation-sample/a/population-and-sample-standard-deviation-review#:~:text=Here's%20how%20to%20calculate%20sample,mean%20from%20each%20data%20point.&text=Step%205%3A%20Divide%20the%20sum%20by%20one%20less%20than%20the,data%20points%20in%20the%20sample
[3.4] Standard deviation calculation; Exceljet; https://exceljet.net/formula/standard-deviation-calculation
[3.5] Difference between STDEVPA and STDEVP functions; Microsoft support; https://docs.microsoft.com/en-us/office/troubleshoot/excel/statistical-functions-differences
[3.6] Standard Deviation Functions; help.gooddata; 
https://help.gooddata.com/doc/en/ reporting-and-dashboards/maql-analytical-query-language/maql-expression-reference/aggregation-functions/statistical-functions/standard-deviation-functions
